ledger/complete/mtrie/README.md

```diff
@@ -113,22 +116,213 @@ On a high level, `MTrie` has the following optimizations:
-    - a `key` and a `value`; 
+    - a payload with a `key` and a `value`;
+    - a path derived from the payload key. 
     - a height value `h`, which can be zero or larger
     - its hash is: subtree-root hash for a tree that only holds `key` and `value`. 
-      To compute this hash, we essentially start with `H(key, value)` and hash our way 
+      To compute this hash, we essentially start with `H(path, value)` and hash our way 
       upwards the tree until we hit height `h`. While climbing the tree upwards, 
       we use the respective `defaultHash[..]` for the other branch which we are merging with. 
    
 Furthermore, an `MTrie` 
 * uses `SHA3-256` as the hash function `H`
-* the registers have keys with `len(key) = 8*l [bits]`, for `l` the key size in bytes
+* the registers have paths with `len(path) = 8*l [bits]`, for `l` the path size in bytes.
 * the height of `MTrie` (per definition, the `height` of the root node) is also `8*l`,
-  for `l` the key size in bytes  
+  for `l` the path size in bytes.  
+* l is fixed to 32 in the current implementation, which makes paths be 256-bits long 
+and the trie root at a height 256.
+  
+### The Mtrie Update algorithm:
 
+Updating register payloads of the mtrie is implemented through copy-on-write, 
+which creates a new `MTrie`, i.e. a new snapshot of the updated key-value store.
+For minimal memory consumption, all sub-tries that were not affected by the update 
+operation are shared between the original `MTrie` and the updated `MTrie`. This means 
+children of some new nodes of the new `Mtrie` point to existing sub-tries from the 
+original `Mtrie`.
 
+The update algorithm `NewTrieWithUpdates` takes a trie `m` as input along with a set of `K` 
+pairs: `(paths[k], payloads[k])` where `k = 0, 1, ..., K-1`.
+It outputs a new trie `new_m` such that each payload `payloads[k]` is stored at the path `paths[k]`.  
+Any path that is not included in the input pairs keeps the payload from the original input `m`.
+
+We first describe the algorithm to perform only a single register update `(path, payload)` 
+and subsequently generalize to an arbitrary number of `K` register updates. 
+Given a root node of a trie, a path and a payload,
+we look for the register addressed by the given path in a recursive top-down manner. Each recursive step 
+operates at a certain height of the tree, starting from the root. Looking at the respective bit 
+`path[i]` (with `i = 256 - height`) of the path, we recursively descend into the left or right child to apply the update.
+For each node visited on the recursive descent, we create a new node at the respective height to represent the updated 
+sub-trie. If the sub-trie is not affected by the update, we re-use the same visited node. 
+
+We define the function `Update` to implement the recursive algorithm to apply the register update.
+`Update` takes as inputs:
+* the `node`, up to which the recursion has progressed in the existing trie (before the update). 
```

Comment:
```suggestion
* `node` is the vertex of the trie before the register update. The `Update` method should return `node` if there are no updates in the respective sub-trie (with root `node`), to avoid unnecessary data duplication. If `node` is `nil`, then there is no candidate in the trie before the register update that could be returned and a new node must necessarily be created.
```